File: integrations/spark/spark-shell-client/spark-shell-client-app/src/main/scala/com/thinkbiganalytics/spark/metadata/TransformJob.scala
package com.thinkbiganalytics.spark.metadata

import com.thinkbiganalytics.spark.rest.model.TransformResponse

import org.apache.spark.SparkContext
import org.apache.spark.scheduler.StageInfo

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.{Callable, FutureTask}

/** A Spark transformation job.
  *
  * This class is thread-safe but `run` should only be invoked once.
  *
  * @param groupId      the group id
  * @param callable     the transformation function
  * @param sparkContext the Spark context
  */
class TransformJob(val groupId: String, private val callable: Callable[TransformResponse], private val sparkContext: SparkContext) extends FutureTask[TransformResponse](callable) {

    /** Spark job id */
    var jobId = Option(0)

    /** List of Spark stages */
    private var _stages: Seq[StageInfo] = Seq()

    /** Number of completed tasks */
    private val tasksCompleted: AtomicInteger = new AtomicInteger(0)

    /** Total number of tasks */
    private val tasksTotal: AtomicInteger = new AtomicInteger(0)

    /** Sets the job progress to completed. */
    def onJobEnd(): Unit = {
        tasksCompleted.set(tasksTotal.get())
    }

    /** Sets the job progress for the specified stage.
      *
      * @param stage          the current stage
      * @param tasksCompleted the number of completed tasks
      */
    def onStageProgress(stage: StageInfo, tasksCompleted: Int): Unit = {
        this.tasksCompleted.set(stages.filter(_.stageId < stage.stageId).map(_.numTasks).sum + tasksCompleted)
    }

    /** Sets a task progress to completed. */
    def onTaskEnd(): Unit = {
        tasksCompleted.incrementAndGet()
    }

    /** Gets the progress of this transformation
      *
      * @return the progress from 0.0 to 1.0
      */
    def progress: Double = {
        val total = tasksTotal.get()
        if (total > 0) {
            tasksCompleted.get().toDouble / total
        } else {
            0.0
        }
    }

    override def run() {
        sparkContext.setJobGroup(groupId, "Transform Job", interruptOnCancel = false)
        super.run()
        sparkContext.clearJobGroup
    }

    override protected def runAndReset: Boolean = {
        throw new UnsupportedOperationException
    }

    /** Gets the Spark stages. */
    def stages = _stages

    /** Sets the Spark stages. */
    def stages_=(stages: Seq[StageInfo]) = {
        _stages = stages
        tasksCompleted.set(0)
        tasksTotal.set(stages.map(_.numTasks).sum)
    }
}

File: integrations/spark/spark-shell-client/spark-shell-client-app/src/main/java/com/thinkbiganalytics/spark/metadata/TransformJob.java
package com.thinkbiganalytics.spark.metadata;

/*-
 * #%L
 * kylo-spark-shell-client-app
 * %%
 * Copyright (C) 2017 ThinkBig Analytics
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.google.common.base.Supplier;
import com.thinkbiganalytics.spark.rest.model.TransformResponse;

import org.apache.spark.SparkContext;

import javax.annotation.Nonnull;

/**
 * A Spark transformation job.
 *
 * <p>This class is thread-safe but {@link #run()} should only be invoked once.</p>
 */
public class TransformJob extends Job<TransformResponse> {

    /**
     * Constructs a {@code TransformJob}.
     */
    public TransformJob(@Nonnull final String groupId, @Nonnull final Supplier<TransformResponse> supplier, @Nonnull final SparkContext sparkContext) {
        super(groupId, supplier, sparkContext);
    }
}
